styx-common/src/test/java/com/spotify/styx/util/CachedSupplierTest.java

```diff
@@ -20,99 +20,196 @@
-    when(delegate.get()).then(a -> x);
-    sut = new CachedSupplier<>(delegate, () -> instant, 10_000);
+  public void setUp() throws Throwable {
+    when(delegate.get()).thenReturn(INITIAL_VALUE);
+    when(time.nanoTime()).then(a -> MILLISECONDS.toNanos(now.toEpochMilli()));
+    sut = new CachedSupplier<>(delegate, time, TIMEOUT);
   }
 
   @Test
-  public void testCachesCalls() throws Exception {
-    int a = sut.get();
-    x = 100;
-    int b = sut.get();
+  public void testCachesCalls() throws Throwable {
+    // Initial load
+    var a = sut.get();
+    assertThat(a, is(INITIAL_VALUE));
+    verify(delegate, times(1)).get();
+
+    // Make underlying supplier return new value
+    lenient().when(delegate.get()).thenReturn(NEW_VALUE);
+
+    // Should still get initial value here
+    var b = sut.get();
+    assertThat(b, is(INITIAL_VALUE));
+
+    // Verify there were no additional calls to the underlying supplier
+    verifyNoMoreInteractions(delegate);
+  }
 
+  @Test
+  public void testReturnsOldValueIfDelegateFails() throws Throwable {
+    // Initial load
+    var a = sut.get();
+    assertThat(a, is(INITIAL_VALUE));
     verify(delegate, times(1)).get();
 
-    assertThat(a, is(42));
-    assertThat(b, is(42));
+    // Make the supplier fail and progress time enough to trigger refresh
+    when(delegate.get()).thenThrow(new RuntimeException("Fail!"));
+    now = now.plus(TIMEOUT).plusSeconds(1);
+
+    // Trigger refresh
+    var b = sut.get();
+    assertThat(b, is(INITIAL_VALUE));
+
+    // Ensure that the old value is still returned (never observe another value)
+    exception.expect(ConditionTimeoutException.class);
+    await().atMost(5, SECONDS).until(() -> sut.get() != INITIAL_VALUE);
   }
 
   @Test
-  public void testCacheTimesOut() throws Exception {
-    int a = sut.get();
-    x = 100;
-    instant = Instant.parse("2016-10-17T15:00:11Z");
-    int b = sut.get();
+  public void testCacheRefreshes() throws Throwable {
+    // Initial load
+    var a = sut.get();
+    assertThat(a, is(INITIAL_VALUE));
+    verify(delegate, times(1)).get();
+
+    // Make supplier return new value and progress time enough to trigger refresh
+    when(delegate.get()).thenReturn(NEW_VALUE);
+    now = now.plus(TIMEOUT).plusSeconds(1);
 
-    verify(delegate, times(2)).get();
-    assertThat(a, is(42));
-    assertThat(b, is(100));
+    // Eventually observe new value
+    await().until(() -> sut.get() == NEW_VALUE);
   }
 
   @Test
-  public void testCacheDoesNotTimeOutWithinTimeout() throws Exception {
+  public void testCacheDoesNotRefreshBeforeTimeout() throws Throwable {
+    // Initial load
     int a = sut.get();
-    x = 100;
-    instant = Instant.parse("2016-10-17T15:00:09Z");
+    assertThat(a, is(INITIAL_VALUE));
+    verify(delegate, times(1)).get();
+
+    // Make supplier return new value and progress time, but not enough to trigger refresh
+    lenient().when(delegate.get()).thenReturn(NEW_VALUE);
```

Comment:
And here.

Reply:
Same. And removing these makes the tests more fragile in the face of future changes.